packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol

```diff
@@ -0,0 +1,171 @@
+        assertEq(minters[0].balance, INITIAL_BALANCE - MINT_FEE);
+    }
+
+    function test_mint_admin() public {
+        vm.prank(admin);
+        token.mint(minters[1]);
+        assertEq(token.totalSupply(), 1);
+        assertEq(token.ownerOf(0), minters[1]);
+    }
+
+    function test_winnerFlow() public {
+        // have all minters mint
+        vm.prank(minters[0]);
+        token.mint{ value: MINT_FEE }();
+        vm.prank(minters[1]);
+        token.mint{ value: MINT_FEE }();
+        vm.prank(minters[2]);
+        token.mint{ value: MINT_FEE }();
+
+        // arbitrarily, minters[0] wins
+        uint256 winnerTokenId = 0;
+
+        // set minters[0] as winner
+        vm.prank(admin);
+        uint256[] memory winners = new uint256[](1);
+        winners[0] = winnerTokenId;
+        token.setWinners(winners);
+        // check winner with both tokenId and address
+        assertTrue(token.isWinner(winnerTokenId));
+        assertTrue(token.isWinner(minters[0]));
+        // and the contract is paused
+        assertTrue(token.paused());
+        // ensure the contract's balance
+        assertEq(address(token).balance, MINT_FEE * minters.length);
+    }
+
+    function test_withdraw() public {
+        test_winnerFlow();
+        uint256 collectedEth = address(token).balance;
+        vm.prank(admin);
+        token.withdraw();
+        assertEq(payoutWallet.balance, collectedEth);
+        assertEq(address(token).balance, 0);
+    }
+
+    function test_ipfs_metadata() public {
+        // ensure URIs are "ticket" before setting winners
+        assertEq(token.baseURI(), "ipfs://baseURI");
+        assertEq(token.tokenURI(0), "ipfs://baseURI/raffle.json");
+        assertEq(token.tokenURI(1), "ipfs://baseURI/raffle.json");
+        // run winner flow
+        test_winnerFlow();
+        // ensure URIs are "winner" and "loser" after setting winners
+        assertEq(token.tokenURI(0), "ipfs://baseURI/winner.json");
+        assertEq(token.tokenURI(1), "ipfs://baseURI/loser.json");
+    }
+
+    function test_upgrade() public {
+        // create the v1 instance of the token
+        vm.startBroadcast(admin);
+
+        address impl = address(new TaikoPartyTicket());
+        address proxy = address(
+            new ERC1967Proxy(
+                impl,
+                abi.encodeCall(
+                    TaikoPartyTicket.initialize,
+                    (payoutWallet, MINT_FEE, "ipfs://baseURI", blacklist)
+                )
+            )
+        );
+
+        TaikoPartyTicket tokenV1 = TaikoPartyTicket(proxy);
+        vm.stopBroadcast();
+
+        // mint some tokens
+        vm.prank(minters[0]);
+        tokenV1.mint{ value: MINT_FEE }();
+        vm.prank(minters[1]);
+        tokenV1.mint{ value: MINT_FEE }();
+        vm.prank(minters[2]);
+        tokenV1.mint{ value: MINT_FEE }();
+
+        // upgrade to v2
+        vm.startBroadcast(admin);
+
+        tokenV1.upgradeToAndCall(
```

Comment:
try to prank a non-admin, to verify the upgradeToAndCall revert, then prank the admin to see the same tx succeed